app/assets/javascripts/discourse/app/components/user-preferences/user-passkeys.gjs

```diff
@@ -0,0 +1,252 @@
+        challenge: Uint8Array.from(response.challenge, (c) => c.charCodeAt(0)),
+        rp: {
+          name: response.rp_name,
+          id: response.rp_id,
+        },
+        user: {
+          id: Uint8Array.from(response.user_secure_id, (c) => c.charCodeAt(0)),
+          name: this.currentUser.username,
+          displayName: this.currentUser.username,
+        },
+        pubKeyCredParams: response.supported_algorithms.map((alg) => {
+          return { type: "public-key", alg };
+        }),
+        excludeCredentials: response.existing_passkey_credential_ids.map(
+          (credentialId) => {
+            return {
+              type: "public-key",
+              id: stringToBuffer(atob(credentialId)),
+            };
+          }
+        ),
+        authenticatorSelection: {
+          // https://www.w3.org/TR/webauthn-2/#user-verification
+          // for passkeys (first factor), user verification should be marked as required
+          // it ensures browser prompts user for PIN/fingerprint/faceID before authenticating
+          userVerification: "required",
+          // See https://w3c.github.io/webauthn/#sctn-createCredential for context
+          // This ensures that the authenticator stores a client-side private key
+          // physical security keys (like Yubikey) need this
+          requireResidentKey: true,
+        },
+      };
+
+      const credential = await navigator.credentials.create({
+        publicKey: publicKeyCredentialCreationOptions,
+      });
+
+      let credentialParam = {
+        id: credential.id,
+        rawId: bufferToBase64(credential.rawId),
+        type: credential.type,
+        attestation: bufferToBase64(credential.response.attestationObject),
+        clientData: bufferToBase64(credential.response.clientDataJSON),
+        name: this.passkeyDefaultName(),
+      };
+
+      const registrationResponse = await this.args.model.registerPasskey(
+        credentialParam
+      );
+
+      if (registrationResponse.error) {
+        this.dialog.alert(registrationResponse.error);
+        return;
+      }
+
+      this.router.refresh();
+
+      // Prompt to rename key after creating
+      this.dialog.dialog({
+        title: I18n.t("user.passkeys.passkey_successfully_created"),
+        type: "notice",
+        bodyComponent: RenamePasskey,
+        bodyComponentModel: registrationResponse,
+      });
+    } catch (error) {
+      this.errorMessage =
+        error.name === "InvalidStateError"
+          ? I18n.t("user.passkeys.already_added_error")
+          : I18n.t("user.passkeys.not_allowed_error");
+      this.dialog.alert(this.errorMessage);
+    }
+  }
+
+  confirmDelete(id) {
+    schedule("afterRender", () => {
+      this.dialog.deleteConfirm({
+        title: I18n.t("user.passkeys.confirm_delete_passkey"),
+        didConfirm: () => {
+          this.args.model.deletePasskey(id).then(() => {
+            this.router.refresh();
+          });
+        },
```

Comment:
There are various cases like this where we have promises which are kinda lost, ultimately it can make the UI complicated to reason about, for example here you need some `schedule("afterRender"` which should really be necessary from what I see